core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java

```diff
@@ -827,39 +819,71 @@ public boolean isActive() throws UnknownTransactionStateException, RepositoryExc
 		}
 	}
 
+	@Override
+	protected void addWithoutCommit(Statement st, Resource... contexts)
+			throws RepositoryException {
+		flushPendingRemoves();
+		if (contexts.length == 0) {
+			pendingAdds.add(st);
+		} else {
+			pendingAdds.add(st.getSubject(), st.getPredicate(), st.getObject(), contexts);
+		}
+	}
+
 	@Override
 	protected void addWithoutCommit(Resource subject, IRI predicate, Value object, Resource... contexts)
 			throws RepositoryException {
-		ValueFactory f = getValueFactory();
+		flushPendingRemoves();
+		pendingAdds.add(subject, predicate, object, contexts);
+	}
 
-		Statement st = f.createStatement(subject, predicate, object);
+	private void flushPendingRemoves() {
+		if (!pendingRemoves.isEmpty()) {
+			for (Resource context : pendingRemoves.contexts()) {
+				String sparqlCommand = createDeleteDataCommand(pendingRemoves.getStatements(null, null, null, context),
+						context);
+				sparqlTransaction.append(sparqlCommand);
+				sparqlTransaction.append("; ");
+			}
+			pendingRemoves.clear();
+		}
+	}
 
-		List<Statement> list = new ArrayList<>(1);
-		list.add(st);
-		String sparqlCommand = createInsertDataCommand(list, contexts);
+	private void flushPendingAdds() {
+		if (!pendingAdds.isEmpty()) {
+			for (Resource context : pendingAdds.contexts()) {
+				String sparqlCommand = createInsertDataCommand(pendingAdds.getStatements(null, null, null, context),
+						context);
+				sparqlTransaction.append(sparqlCommand);
+				sparqlTransaction.append("; ");
+			}
+			pendingAdds.clear();
+		}
+	}
 
-		sparqlTransaction.append(sparqlCommand);
-		sparqlTransaction.append("; ");
+	@Override
+	protected void removeWithoutCommit(Statement st, Resource... contexts) throws RepositoryException {
+		flushPendingAdds();
+		if (contexts.length == 0) {
+			pendingRemoves.add(st);
+		} else {
+			pendingRemoves.add(st.getSubject(), st.getPredicate(), st.getObject(), contexts);
+		}
 	}
 
 	@Override
 	protected void removeWithoutCommit(Resource subject, IRI predicate, Value object, Resource... contexts)
 			throws RepositoryException {
-		String sparqlCommand = "";
-		if (subject != null && predicate != null && object != null) {
-			ValueFactory f = getValueFactory();
-
-			Statement st = f.createStatement(subject, predicate, object);
+		flushPendingAdds();
 
-			List<Statement> list = new ArrayList<>(1);
-			list.add(st);
-			sparqlCommand = createDeleteDataCommand(list, contexts);
+		if (subject != null && predicate != null && object != null) {
+			pendingRemoves.add(subject, predicate, object, contexts);
```

Comment:
Bit hazy in the morning but I wonder how you handle removing from all contexts (implicitly).

Reply:
Ah - good call. I'll write a test case for that.

Reply:
Just to be clear: this is in fact handled by just mapping to a DELETE DATA operation without a GRAPH clause. Funnily enough the integration tests didn't cover this specific case, so I've added a few unit tests as well as a new integration test.